src/react/hooks/useQuery.ts

```diff
@@ -1,16 +1,355 @@
+
+import { DocumentType, verifyDocumentType } from '../parser';
+import { useApolloClient } from './useApolloClient';
 
-export function useQuery<TData = any, TVariables = OperationVariables>(
+export function useQuery<
+  TData = any,
+  TVariables = OperationVariables,
+>(
   query: DocumentNode | TypedDocumentNode<TData, TVariables>,
-  options?: QueryHookOptions<TData, TVariables>
-) {
-  return useBaseQuery<TData, TVariables>(query, options, false) as QueryResult<
-    TData,
-    TVariables
-  >;
+  options?: QueryHookOptions<TData, TVariables>,
+): QueryResult<TData> {
+  const context = useContext(getApolloContext());
+  const client = useApolloClient(options?.client);
+  verifyDocumentType(query, DocumentType.Query);
+  const [obsQuery, setObsQuery] = useState(() => {
+    const watchQueryOptions = createWatchQueryOptions(query, options);
+    // See if there is an existing observable that was used to fetch the same
+    // data and if so, use it instead since it will contain the proper queryId
+    // to fetch the result set. This is used during SSR.
+    let obsQuery: ObservableQuery<TData, TVariables> | null = null;
+    if (context.renderPromises) {
+      obsQuery = context.renderPromises.getSSRObservable(watchQueryOptions);
+    }
+
+    if (!obsQuery) {
+      // Is it safe (StrictMode/memory-wise) to call client.watchQuery here?
+      obsQuery = client.watchQuery(watchQueryOptions);
+      if (context.renderPromises) {
+        context.renderPromises.registerSSRObservable(
+          obsQuery,
+          watchQueryOptions,
+        );
+      }
+    }
+
+    if (
+      context.renderPromises &&
+      options?.ssr !== false &&
+      !options?.skip &&
+      obsQuery.getCurrentResult().loading
+    ) {
+      // TODO: This is a legacy API which could probably be cleaned up
+      context.renderPromises.addQueryPromise(
+        {
+          // The only options which seem to actually be used by the
+          // RenderPromises class are query and variables.
+          getOptions: () => createWatchQueryOptions(query, options),
+          fetchData: () => new Promise<void>((resolve) => {
+            const sub = obsQuery!.subscribe({
+              next(result) {
+                if (!result.loading) {
+                  resolve()
+                  sub.unsubscribe();
+                }
+              },
+              error() {
+                resolve();
+                sub.unsubscribe();
+              },
+              complete() {
+                resolve();
+              },
+            });
+          }),
+        },
+        // This callback never seemed to do anything
+        () => null,
+      );
+    }
+
+    return obsQuery;
+  });
+
+  let [result, setResult] = useState(() => {
+    const result = obsQuery.getCurrentResult();
+    if (!result.loading) {
+      if (result.data) {
+        options?.onCompleted?.(result.data);
+      } else if (result.error) {
+        options?.onError?.(result.error);
+      }
```

Comment:
```suggestion
    if (!result.loading && options) {
      if (result.error) {
        options.onError?.(result.error);
      } else if (result.data) {
        options.onCompleted?.(result.data);
      }
```
If I'm reading the [`onCompleted` and `onError` docs](https://www.apollographql.com/docs/react/data/queries/#oncompleted) correctly, we should give `result.error` preference here, and in the couple of other places where we call `onCompleted` and `onError`.

Reply:
Done. I think it might be nice to streamline the result modification and callback calling code somehow, but haven’t figured out where to do it. For instance, we call onError in the observable onError callback and onCompleted in the observable onNext callbacks, but should these methods be called someplace else?

Reply:
Do we have a `Promise<ApolloQueryResult<TData>>` somewhere that we could attach `onCompleted` and/or `onError` to, as `then` handlers?

Reply:
Nope. The difficult part is that an “ApolloQueryResult” is created in multiple places, in the initial useState callback, in the observable callbacks, and based on whether or not the query is skipped. Trying to abstract over all these cases is hard, especially because any function we define inline might close over stale variables.